lib/site-inspector.rb

```diff
@@ -127,41 +190,447 @@ def redirect
+
+    www = !!(
+      (
+        combos[:https][:www][:up] or
+        combos[:http][:www][:up]
+      ) and (
+        (
+          combos[:https][:root][:redirect] or
+          !combos[:https][:root][:up] or
+          combos[:https][:root][:https_bad_name] or
+          !combos[:https][:root][:status].to_s.start_with?("2")
+        ) and (
+          combos[:http][:root][:redirect] or
+          !combos[:http][:root][:up] or
+          !combos[:http][:root][:status].to_s.start_with?("2")
+        )
+      ) and (
+        (
+          (
+            !combos[:https][:root][:up] or
+            combos[:https][:root][:https_bad_name] or
+            !combos[:https][:root][:status].to_s.start_with?("2")
+          ) and
+          (
+            !combos[:http][:root][:up] or
+            !combos[:http][:root][:status].to_s.start_with?("2")
+          )
+        ) or
+        (
+          combos[:https][:root][:redirect_immediately_to_www] and
+          !combos[:https][:root][:redirect_immediately_external]
+        ) or
+        (
+          combos[:http][:root][:redirect_immediately_to_www] and
+          !combos[:http][:root][:redirect_immediately_external]
+        )
+      )
+    )
+
+    # A domain is "canonically" at https if:
+    #  * at least one of its https endpoints is live and
+    #    doesn't have an invalid hostname
+    #  * both http endpoints are either down or redirect *somewhere*
+    #  * at least one http endpoint redirects immediately to
+    #    an *internal* https endpoint
+    # This is meant to affirm situations like:
+    #   http:// -> http://www -> https://
+    #   https:// -> http:// -> https://www
+    # and meant to avoid affirming situations like:
+    #   http:// -> http://non-www
+    #   http://www -> http://non-www
+    # or:
+    #   http:// -> 200, http://www -> https://www
+    #
+    # It allows a site to be canonically HTTPS if the cert has
+    # a valid hostname but invalid chain issues.
+
+    https = !!(
+      (
+        (
+          combos[:https][:root][:up] and
+          !combos[:https][:root][:https_bad_name]
+        ) or
+        (
+          combos[:https][:www][:up] and
+          !combos[:https][:www][:https_bad_name]
+        )
+      ) and (
+        (
+          combos[:http][:root][:redirect] or
+          !combos[:http][:root][:up] or
+          !combos[:http][:root][:status].to_s.start_with?("2")
+        ) and (
+          combos[:http][:www][:redirect] or
+          !combos[:http][:www][:up] or
+          !combos[:http][:www][:status].to_s.start_with?("2")
+        )
+      ) and (
+        (
+          combos[:http][:root][:redirect_immediately_to_https] and
+          !combos[:http][:root][:redirect_immediately_external]
+        ) or (
+          combos[:http][:www][:redirect_immediately_to_https] and
+          !combos[:http][:www][:redirect_immediately_external]
+        )
+      )
+    )
+
+    details[:canonical_endpoint] = www ? :www : :root
```

Comment:
I'd make each one of the "details" a method. The `http` method should simply build a hash. It shouldn't perform any calculations itself.